src/app/qgsrelationmanagerdialog.cpp

```diff
@@ -15,37 +15,79 @@
 #include "qgsrelationadddlg.h"
+#include "qgsrelationaddpolymorphicdlg.h"
 #include "qgsrelationmanagerdialog.h"
 #include "qgsrelationmanager.h"
+#include "qgspolymorphicrelation.h"
 #include "qgsvectorlayer.h"
 
+
+#ifndef SIP_RUN
+class RelationNameEditorDelegate: public QStyledItemDelegate
+{
+  public:
+    RelationNameEditorDelegate( const QList<int> &editableColumns, QObject *parent = nullptr )
+      : QStyledItemDelegate( parent )
+      , mEditableColumns( editableColumns )
+    {}
+
+    virtual QWidget *createEditor( QWidget *parentWidget, const QStyleOptionViewItem &option, const QModelIndex &index ) const
+    {
+      if ( mEditableColumns.contains( index.column() ) )
+        return QStyledItemDelegate::createEditor( parentWidget, option, index );
+
+      return nullptr;
+    }
+  private:
+    QList<int> mEditableColumns;
+};
+#endif
+
 QgsRelationManagerDialog::QgsRelationManagerDialog( QgsRelationManager *relationMgr, QWidget *parent )
   : QWidget( parent )
   , Ui::QgsRelationManagerDialogBase()
   , mRelationManager( relationMgr )
 {
   setupUi( this );
+
+  mRelationsTree->header()->setSectionResizeMode( QHeaderView::ResizeToContents );
+  mRelationsTree->setItemDelegate( new RelationNameEditorDelegate( QList<int>( {0} ), mRelationsTree ) );
+
   connect( mBtnAddRelation, &QPushButton::clicked, this, &QgsRelationManagerDialog::mBtnAddRelation_clicked );
+  connect( mActionAddPolymorphicRelation, &QAction::triggered, this, &QgsRelationManagerDialog::mActionAddPolymorphicRelation_triggered );
+  connect( mActionEditPolymorphicRelation, &QAction::triggered, this, &QgsRelationManagerDialog::mActionEditPolymorphicRelation_triggered );
   connect( mBtnDiscoverRelations, &QPushButton::clicked, this, &QgsRelationManagerDialog::mBtnDiscoverRelations_clicked );
   connect( mBtnRemoveRelation, &QPushButton::clicked, this, &QgsRelationManagerDialog::mBtnRemoveRelation_clicked );
 
   mBtnRemoveRelation->setEnabled( false );
-  connect( mRelationsTable->selectionModel(), &QItemSelectionModel::selectionChanged, this, &QgsRelationManagerDialog::onSelectionChanged );
+  mBtnAddRelation->setPopupMode( QToolButton::MenuButtonPopup );
+  mBtnAddRelation->addAction( mActionAddPolymorphicRelation );
+  mBtnAddRelation->addAction( mActionEditPolymorphicRelation );
+
+  connect( mRelationsTree->selectionModel(), &QItemSelectionModel::selectionChanged, this, &QgsRelationManagerDialog::onSelectionChanged );
 }
 
 void QgsRelationManagerDialog::setLayers( const QList< QgsVectorLayer * > &layers )
 {
   mLayers = layers;
 
   const QList<QgsRelation> &relations = mRelationManager->relations().values();
-
-  const auto constRelations = relations;
-  for ( const QgsRelation &rel : constRelations )
+  for ( const QgsRelation &rel : relations )
   {
+    // the generated relations for polymorphic relations should be ignored,
+    // they are generated when the polymorphic relation is added to the table
+    if ( !rel.polymorphicRelationId().isEmpty() )
```

Comment:
for the sake of readability, should we have `bool QgsRelation::isPolymorphic() const`

Reply:
I would prefer to have `QgsRelation::RelationType QgsRelation::type() const`, it is more future proof if there are other types of relations added. The two types for now would be `RelationType::Normal` and `RelationType::Generated`.

On your proposal, I am not sure if we should use the same term on `Polymorphic` for `QgsRelation`, since we already have `QgsPolymorphicRelation` and would become a mess. I call them "generated [from a polymorphic relation".